README.md
# TotemDismiss

NOTE - The Phase 2.5 patch broke this addon, as well as many other totem-related addons. See [#1](https://github.com/nmonterroso/TotemDismiss/issues/1).


CHANGELOG.txt
2019.10.23:
- Adds dismiss all. This is enabled by default as a button that shows up before the regular totem buttons. Can be moved after the regular totem buttons, or hidden
- Adds option to hide regular totem buttons
- Info in Advanced options about creating a macro to dismiss all totems (requires this addon), as well as a button that will create a character-specific macro for you to dismiss your totems.

2019.10.22:
- First release!
- lock/unlock
- scale
- cooldown font options

main.lua
TotemDismiss = LibStub("AceAddon-3.0"):NewAddon("TotemDismiss", "AceConsole-3.0", "AceEvent-3.0")
TotemDismiss.dismissAllTotem = 0
TotemDismiss.fireTotem = 1
TotemDismiss.earthTotem = 2
TotemDismiss.waterTotem = 3
TotemDismiss.airTotem = 4
TotemDismiss.dismissAllOrdering = {
  never = 'never',
  before = 'before',
  after = 'after',
}

function TotemDismiss:OnInitialize()
  if not self:PlayerIsShaman() then
    return
  end

  self.db = LibStub("AceDB-3.0"):New("TotemDismissDB", self.defaultVariables)
  self.optionFrame = self:SetupOptions()
  self.config = nil
  self.container = nil
  self.totems = {
    [self.dismissAllTotem] = {
      id = self.dismissAllTotem,
      name = "dismissAll",
      icon = 'Interface/ICONS/Spell_Totem_WardOfDraining',
      button = nil,
    },
    [self.fireTotem] = {
      id = self.fireTotem,
      name = "fire",
      icon = 'Interface/ICONS/Spell_Fire_Fire',
      button = nil,
    },
    [self.earthTotem] = {
      id = self.earthTotem,
      name = "earth",
      icon = 'Interface/ICONS/INV_Stone_16',
      button = nil,
    },
    [self.waterTotem] = {
      id = self.waterTotem,
      name = "water",
      icon = 'Interface/ICONS/INV_Stone_02',
      button = nil,
    },
    [self.airTotem] = {
      id = self.airTotem,
      name = "air",
      icon = 'Interface/AddOns/TotemDismiss/air',
      button = nil,
    },
  }
end

function TotemDismiss:OnEnable()
  if not self:PlayerIsShaman() then
    return
  end

  self.config = self.db.global
  self:Draw()
  self:RegisterEvent("PLAYER_TOTEM_UPDATE", function(event, id)
    self:OnTotemUpdate(id)
    self:EnableOrDisableDismissAll()
  end)
  self:RegisterEvent("PLAYER_LOGOUT", function()
    self.config.locked = true
  end)
end

function TotemDismiss:Draw()
  self.container = self:CreateContainer()
  self:CreateTotems()
  self:EnableOrDisableDismissAll()
end

function TotemDismiss:OnTotemUpdate(id)
  local haveTotem, totemName, startTime, duration, icon = GetTotemInfo(id)
  local totem = self.totems[id]
  if totem.button == nil then
    return
  end

  if haveTotem and startTime > 0 then
    self:EnableButton(totem)
    self:SetButtonTexture(totem, icon)
    totem.button.cooldown:SetCooldown(startTime, duration)
  else
    self:DisableButton(totem)
    self:SetButtonTexture(totem)
  end
end

function TotemDismiss:CreateTotems()
  local lastAnchor
  for _, id in ipairs(self:GetTotemOrder()) do
    local totem = self.totems[id]

    lastAnchor = self:InitTotem(totem, lastAnchor)
    self:OnTotemUpdate(totem.id)
  end
end

function TotemDismiss:InitTotem(totem, anchor)
  totem.button = self:CreateButton(totem, anchor)

  self:SetButtonTexture(totem)
  self:DisableButton(totem)

  if self:IsShowingTotem(totem) then
    return totem.button
  end

  return nil
end

function TotemDismiss:PlayerIsShaman()
  local _, classFilename = UnitClass("player")
  return classFilename == "SHAMAN"
end

function TotemDismiss:serializeTable(val, name, skipnewlines, depth)
  skipnewlines = skipnewlines or false
  depth = depth or 0

  local tmp = string.rep(" ", depth)

  if name then tmp = tmp .. name .. " = " end

  if type(val) == "table" then
    tmp = tmp .. "{" .. (not skipnewlines and "\n" or "")

    for k, v in pairs(val) do
      tmp =  tmp .. self:serializeTable(v, k, skipnewlines, depth + 1) .. "," .. (not skipnewlines and "\n" or "")
    end

    tmp = tmp .. string.rep(" ", depth) .. "}"
  elseif type(val) == "number" then
    tmp = tmp .. tostring(val)
  elseif type(val) == "string" then
    tmp = tmp .. string.format("%q", val)
  elseif type(val) == "boolean" then
    tmp = tmp .. (val and "true" or "false")
  else
    tmp = tmp .. "\"[inserializeable datatype:" .. type(val) .. "]\""
  end

  return tmp
end


config.lua
TotemDismiss.defaultVariables = {
  global = {
    buttonWidth = 32,
    buttonHeight = 32,
    locked = true,
    scale = 1.0,
    margin = 0,
    font = {
      name = "Friz Quadrata TT",
      path = "Fonts\\FRIZQT__.TTF",
      size = 16,
      flags = {
        outline = false,
        thickoutline = false,
        monochrome = false,
      },
    },
    anchor = {
      point = "LEFT",
      relativePoint = "CENTER",
      offsetX = nil,
      offsetY = nil,
    },
    displayTotems = {
      dismissAll = TotemDismiss.dismissAllOrdering.before,
      regular = true,
    },
  }
}

function TotemDismiss:GetContainerPoint()
  local offsetX = -2*self.config.buttonWidth
  if self.config.anchor.offsetX ~= nil then
    offsetX = self.config.anchor.offsetX
  end

  local offsetY = 0
  if self.config.anchor.offsetY ~= nil then
    offsetY = self.config.anchor.offsetY
  end

  return self.config.anchor.point, UIParent, self.config.anchor.relativePoint, offsetX, offsetY
end

function TotemDismiss:GetInitialButtonPoint()
  return "LEFT", self.container, "LEFT"
end

function TotemDismiss:GetButtonPoint(anchor)
  return "LEFT", anchor, "RIGHT", self.config.margin
end

function TotemDismiss:GetButtonWidth()
  return self.config.buttonWidth*self.config.scale
end

function TotemDismiss:GetButtonHeight()
  return self.config.buttonHeight*self.config.scale
end

function TotemDismiss:GetContainerWidth()
  if not self:IsShowingContainer() then
    return 0
  end

  local numButtons = 0
  if self.config.displayTotems.dismissAll ~= 'never' then
    numButtons = numButtons + 1
  end

  if self.config.displayTotems.regular then
    numButtons = numButtons + 4
  end

  return self:GetButtonWidth()*numButtons + self.config.margin*(numButtons - 1)
end

function TotemDismiss:GetFontFlags()
  local flags = {}
  if self.config.font.flags.outline then
    table.insert(flags, "OUTLINE")
  end
  if self.config.font.flags.thickoutline then
    table.insert(flags, "THICKOUTLINE")
  end
  if self.config.font.flags.monochrome then
    table.insert(flags, "MONOCHROME")
  end

  if table.getn(flags) == 0 then
    return nil
  end

  return table.concat(flags, ", ")
end

function TotemDismiss:IsShowingContainer()
  return self.config.displayTotems.dismissAll ~= self.dismissAllOrdering.never or
      self.config.displayTotems.regular
end

function TotemDismiss:IsShowingTotem(totem)
  if totem.id == self.dismissAllTotem then
    return self.config.displayTotems.dismissAll ~= self.dismissAllOrdering.never
  end

  return self.config.displayTotems.regular
end

function TotemDismiss:GetTotemOrder()
  local order = {}
  if self.config.displayTotems.dismissAll ~= self.dismissAllOrdering.after then
    table.insert(order, TotemDismiss.dismissAllTotem)
  end

  table.insert(order, TotemDismiss.earthTotem)
  table.insert(order, TotemDismiss.fireTotem)
  table.insert(order, TotemDismiss.waterTotem)
  table.insert(order, TotemDismiss.airTotem)

  if self.config.displayTotems.dismissAll == self.dismissAllOrdering.after then
    table.insert(order, TotemDismiss.dismissAllTotem)
  end

  return order
end


buttons.lua
function TotemDismiss:CreateButton(totem, anchor)
  local button

  if totem.button ~= nil then
    button = totem.button
  elseif totem.id == self.dismissAllTotem then
    button = self:CreateDismissAllTotemButton(totem)
  else
    button = self:CreateRegularTotemButton(totem)
  end

  if self:IsShowingTotem(totem) then
    if anchor == nil then
      button:SetPoint(self:GetInitialButtonPoint())
    else
      button:SetPoint(self:GetButtonPoint(anchor))
    end

    button:Show()
  else
    button:Hide()
  end

  button:SetFrameStrata("LOW")
  button:SetWidth(self:GetButtonWidth())
  button:SetHeight(self:GetButtonHeight())

  button.texture = self:CreateButtonTexture(button)
  button.overlay = self:CreateButtonOverlay(button)
  button.cooldown = self:CreateButtonCooldown(button)

  button:SetNormalTexture(button.texture)

  return button
end

function TotemDismiss:CreateDismissAllTotemButton(totem)
  local macroText = [[
    /click TotemDismissButton_earth
    /click TotemDismissButton_fire
    /click TotemDismissButton_water
    /click TotemDismissButton_air
  ]]
  local button = CreateFrame("Button", "TotemDismissButton_"..totem.name, self.container, "SecureActionButtonTemplate")
  button:SetAttribute("type1", "macro")
  button:SetAttribute("macrotext", macroText:trim():gsub("%s+", " "):gsub(" /", "\n/"))

  return button
end

function TotemDismiss:CreateRegularTotemButton(totem)
  local button = CreateFrame("Button", "TotemDismissButton_"..totem.name, self.container, "SecureActionButtonTemplate")
  button:SetAttribute("type1", "destroytotem")
  button:SetAttribute("*totem-slot*", totem.id)

  return button
end

function TotemDismiss:CreateButtonOverlay(button)
  local overlay = button.overlay
  if button.overlay == nil then
    overlay = button:CreateTexture(nil, "OVERLAY")
  end

  overlay:SetPoint("TOPLEFT", 1, -1)
  overlay:SetPoint("BOTTOMRIGHT", -1, 1)
  overlay:SetColorTexture(0, 0, 0, .55)

  return overlay
end

function TotemDismiss:CreateButtonTexture(button)
  local texture = button.texture
  if texture == nil then
    texture = button:CreateTexture()
  end

  texture:SetAllPoints()
  return texture
end

function TotemDismiss:CreateButtonCooldown(button)
  local cooldown = button.cooldown
  if cooldown == nil then
    cooldown = CreateFrame("Cooldown", nil, button, "CooldownFrameTemplate")
  end

  cooldown:SetReverse(true)
  cooldown:SetDrawEdge(false)
  cooldown:SetPoint("TOPLEFT", 1, -1)
  cooldown:SetPoint("BOTTOMRIGHT", -1, 1)
  cooldown.text = cooldown:GetRegions()
  cooldown.text:SetFont(self.config.font.path, self.config.font.size, self:GetFontFlags())

  return cooldown
end

function TotemDismiss:SetButtonTexture(totem, icon)
  if totem.button == nil then
    return
  end

  if icon == nil then
    icon = totem.icon
  end

  totem.button.texture:SetTexture(icon)
end

function TotemDismiss:EnableButton(totem)
  if totem.button ~= nil then
    totem.button.overlay:Hide()
  end
end

function TotemDismiss:DisableButton(totem)
  if totem.button ~= nil then
    totem.button.overlay:Show()
    totem.button.cooldown:Clear()
  end
end

function TotemDismiss:EnableButton(totem)
  if totem.button ~= nil then
    totem.button.overlay:Hide()
  end
end

function TotemDismiss:EnableOrDisableDismissAll()
  local dismissAllTotem = self.totems[self.dismissAllTotem]

  for id, _ in pairs(self.totems) do
    if id ~= self.dismissAllTotem then
      local haveTotem, _, startTime, _, _ = GetTotemInfo(id)
      if haveTotem and startTime > 0 then
        self:EnableButton(dismissAllTotem)
        return
      end
    end
  end

  self:DisableButton(dismissAllTotem)
end


container.lua
function TotemDismiss:CreateContainer()
  local container = self.container
  if self.container == nil then
    container = CreateFrame("Frame", nil, mainFrame)
  end

  container:SetClampedToScreen(true)
  container:SetPoint(self:GetContainerPoint())
  container:SetHeight(self:GetButtonHeight())
  container:SetWidth(self:GetContainerWidth())
  container:Show()

  if container.dragTexture == nil then
    local dragTexture = container:CreateTexture()
    dragTexture:SetColorTexture(0, .8, 0, .55)
    dragTexture:SetAllPoints()
    dragTexture:Hide()

    container.dragTexture = dragTexture
  end

  return container
end

function TotemDismiss:UnlockContainer()
  if self.container == nil then
    return
  end

  self.container:SetMovable(true)
  self.container:EnableMouse(true)
  self.container:RegisterForDrag("LeftButton")
  self.container.dragTexture:Show()

  self.container:SetScript("OnDragStart", function()
    self.container:StartMoving()
  end)
  self.container:SetScript("OnDragStop", function()
    self.config.anchor.offsetX = self.container:GetLeft()
    self.config.anchor.offsetY = self.container:GetBottom()
    self.config.anchor.point = "BOTTOMLEFT"
    self.config.anchor.relativePoint = "BOTTOMLEFT"
    self.container:StopMovingOrSizing()
  end)
end

function TotemDismiss:LockContainer()
  if self.container == nil then
    return
  end

  self.container.dragTexture:Hide()
  self.container:SetMovable(false)
  self.container:EnableMouse(false)
  self.container:RegisterForDrag("LeftButton")
end

totem-dismiss.lua
local fire = 'fire'
local earth = 'earth'
local water = 'water'
local air = 'air'

TotemDismiss = LibStub("AceAddon-3.0"):NewAddon(
    "TotemDismiss",
    "AceConsole-3.0",
    "AceEvent-3.0"
    --"AceDB-3.0",
    --"AceConfig-3.0"
)

function TotemDismiss:OnInitialize()
  if not self:canUse() then
    return
  end

  self.db = LibStub("AceDB-3.0"):New("TotemDismissDB")
  self.totemOrder = {earth, fire, water, air}
  self.totems = {
    [fire] = {
      id = 1,
      icon = 'Interface/ICONS/Spell_Fire_Fire',
      button = nil,
      overlay = nil,
      iconTexture = nil,
      cooldown = nil,
      isReady = false,
    },
    [earth] = {
      id = 2,
      icon = 'Interface/ICONS/INV_Stone_16',
      button = nil,
      overlay = nil,
      iconTexture = nil,
      cooldown = nil,
      isReady = false,
    },
    [water] = {
      id = 3,
      icon = 'Interface/ICONS/INV_Stone_02',
      button = nil,
      overlay = nil,
      iconTexture = nil,
      cooldown = nil,
      isReady = false,
    },
    [air] = {
      id = 4,
      icon = 'Interface/AddOns/TotemDismiss/air',
      button = nil,
      overlay = nil,
      iconTexture = nil,
      cooldown = nil,
      isReady = false,
    },
  }
end

function TotemDismiss:OnEnable()
  if not self:canUse() then
    return
  end

  self:initTotems()
  self:RegisterEvent("PLAYER_TOTEM_UPDATE", function(event, id)
    self:onTotemUpdate(id)
  end)
end

function TotemDismiss:onTotemUpdate(id)
  local haveTotem, totemName, startTime, duration, icon = GetTotemInfo(id)
  local totem = self:getTotem(id)
  if not totem.isReady then
    return
  end

  if haveTotem and startTime > 0 then
    self:enable(id)
    self:setTexture(id, icon)
    totem.cooldown:SetCooldown(startTime, duration)
  else
    self:disable(id)
    self:setTexture(id)
  end
end

function TotemDismiss:initTotems()
  local lastAnchor
  for _, type in ipairs(self.totemOrder) do
    local totem = self.totems[type]

    if not self.totems[type].isReady then
      self:initTotem(totem.id, type, lastAnchor)
    end

    if totem.isReady then
      lastAnchor = totem.button
    end
  end
end

function TotemDismiss:initTotem(id, type, anchor)
  local button = self:createButton(id, type, anchor)
  local iconTexture = self:createIconTexture(button)
  local overlay = self:createOverlay(button)
  local cooldown = self:createCooldown(button)

  button:SetNormalTexture(iconTexture)
  button:Show()

  self.totems[type].button = button
  self.totems[type].overlay = overlay
  self.totems[type].iconTexture = iconTexture
  self.totems[type].cooldown = cooldown
  self.totems[type].isReady = true

  self:setTexture(id)
  self:disable(id)
end

function TotemDismiss:setTexture(id, icon)
  local totem = self:getTotem(id)
  if not totem.isReady then
    return
  end

  if icon == nil then
    icon = totem.icon
  end

  totem.iconTexture:SetTexture(icon)
end

function TotemDismiss:getTotem(id)
  if id == 1 then
    return self.totems[fire]
  elseif id == 2 then
    return self.totems[earth]
  elseif id == 3 then
    return self.totems[water]
  end

  return self.totems[air]
end

function TotemDismiss:disable(id)
  local totem = self:getTotem(id)
  if totem.isReady then
    totem.button:Disable()
    totem.overlay:Show()
    totem.cooldown:Clear()
  end
end

function TotemDismiss:enable(id)
  local totem = self:getTotem(id)
  if totem.isReady then
    totem.button:Enable()
    totem.overlay:Hide()
  end
end

function TotemDismiss:createButton(id, type, anchor)
  local button = CreateFrame("Button", "TotemDismissButton_"..type, UIParent, "SecureActionButtonTemplate")

  if anchor == nil then
    button:SetPoint("LEFT", mainFrame, "CENTER", -64, 0)
  else
    button:SetPoint("LEFT", anchor, "RIGHT")
  end

  -- TODO: set strata?
  button:SetWidth(32)
  button:SetHeight(32)
  button:SetAttribute("*type1", "destroytotem")
  button:SetAttribute("*totem-slot*", id)

  return button
end

function TotemDismiss:createCooldown(button)
  local cooldown = CreateFrame("Cooldown", nil, button, "CooldownFrameTemplate")
  cooldown:SetReverse(true)
  cooldown:SetDrawEdge(false)
  cooldown:SetPoint("TOPLEFT", 1, -1)
  cooldown:SetPoint("BOTTOMRIGHT", -1, 1)

  return cooldown
end

function TotemDismiss:createIconTexture(button)
  local texture = button:CreateTexture()
  texture:SetAllPoints()

  return texture
end

function TotemDismiss:createOverlay(button)
  local overlay = button:CreateTexture(nil, "OVERLAY")
  overlay:SetPoint("TOPLEFT", 1, -1)
  overlay:SetPoint("BOTTOMRIGHT", -1, 1)
  overlay:SetColorTexture(0, 0, 0, .55)
  overlay:Show()

  return overlay
end

function TotemDismiss:canUse()
  local _, classFilename = UnitClass("player")
  return classFilename == "SHAMAN"
end


options.lua
local fonts = LibStub("LibSharedMedia-3.0"):HashTable("font")
local macroName = 'TotemDismissAll'
local macroAction = '/click TotemDismissButton_dismissAll'
local options = {
  name = 'TotemDismiss',
  type = "group",
  args = {
    locked = {
      name = "Locked",
      desc = "Prevents the bar from being moved",
      type = "toggle",
      set = function(info, val)
        TotemDismiss.config.locked = val
        if val then
          TotemDismiss:LockContainer()
        else
          TotemDismiss:UnlockContainer()
        end
      end,
      get = function(info) return TotemDismiss.config.locked end
    },
    scale = {
      name = "Scale",
      desc = "Button scale",
      type = "range",
      softMin = 0.1,
      softMax = 1.9,
      set = function(info, val)
        if val < 0.1 then
          val = 0.1
        elseif val > 1.9 then
          val = 1.9
        end

        TotemDismiss.config.scale = val
        TotemDismiss:Draw()
      end,
      get = function(info) return TotemDismiss.config.scale end
    },
    font = {
      name = "Font",
      type = "group",
      args = {
        face = {
          type = "select",
          dialogControl = "LSM30_Font",
          name = "Font Face",
          desc = "Sets the font of the cooldown timer when a totem is active",
          values = LibStub("LibSharedMedia-3.0"):HashTable("font"),
          set = function(info, val)
            TotemDismiss.config.font.name = val
            TotemDismiss.config.font.path = fonts[val]
            TotemDismiss:Draw()
          end,
          get = function(info)
            return TotemDismiss.config.font.name
          end,
        },
        size = {
          name = "Font Size",
          desc = "Size of the cooldown timer when a totem is active",
          type = "range",
          softMin = 1,
          softMax = 32,
          step = 1,
          set = function(info, val)
            if val < 1 then
              val = 1
            elseif val > 32 then
              val = 32
            end

            TotemDismiss.config.font.size = val
            TotemDismiss:Draw()
          end,
          get = function(info)
            return TotemDismiss.config.font.size
          end,
        },
        flags = {
          name = "Other",
          desc = "other modifications to the cooldown font",
          type = "multiselect",
          values = {
            outline = "Outline",
            thickoutline = "Thick Outline",
            monochrome = "Monochrome",
          },
          set = function(info, key, val)
            TotemDismiss.config.font.flags[key] = val
            TotemDismiss:Draw()
          end,
          get = function(info, key)
            return TotemDismiss.config.font.flags[key]
          end,
        },
      },
    },
    advanced = {
      name = "Advanced",
      type = "group",
      order = -1,
      args = {
        showDismissAll = {
          name = "Show Dismiss All Button",
          type = "select",
          desc = "Whether or not to show button to dismiss all totems",
          values = {
            [TotemDismiss.dismissAllOrdering.never] = 'Never',
            [TotemDismiss.dismissAllOrdering.before] = 'Before Regular Totems',
            [TotemDismiss.dismissAllOrdering.after] = 'After Regular Totems',
          },
          set = function(info, val)
            TotemDismiss.config.displayTotems.dismissAll = val
            TotemDismiss:Draw()
          end,
          get = function(info)
            return TotemDismiss.config.displayTotems.dismissAll
          end,
        },
        showRegular = {
          name = "Show Regular",
          type = "toggle",
          desc = "Whether or not to show buttons to dismiss regular totems",
          set = function(info, val)
            TotemDismiss.config.displayTotems.regular = val
            TotemDismiss:Draw()
          end,
          get = function(info)
            return TotemDismiss.config.displayTotems.regular
          end,
        },
        createMacroHeader = {
          name = "Macro-based dismiss",
          type = "header",
          order = -3,
        },
        createMacroDesc = {
          name = "It's possible to create a macro to dismiss all totems. This is recommended if hiding both the dismiss all button and the regular totem buttons. The macro for this is '"..macroAction.."'. Click the button below to have the character specific macro created with the name '"..macroName.."'",
          type = "description",
          order = -2,
        },
        createMacro = {
          name = "Create Macro",
          type = "execute",
          order = -1,
          func = function()
            TotemDismiss:CreateMacro()
          end
        }
      },
    }
  }
}

function TotemDismiss:SetupOptions()
  LibStub("AceConfig-3.0"):RegisterOptionsTable("TotemDismiss", options, {"totemdismiss"})
  local optionFrame = LibStub("AceConfigDialog-3.0"):AddToBlizOptions("TotemDismiss", "Totem Dismiss")
  return optionFrame
end

function TotemDismiss:CreateMacro()
  if UnitAffectingCombat("player") then
    self:Print("Can't create macro when in combat")
    return
  end

  local name = GetMacroInfo(macroName)
  if name ~= nil then
    DeleteMacro(macroName)
  end

  CreateMacro(macroName, 'Spell_Totem_WardOfDraining', macroAction, 1, 1)
  self:Print("Macro created! Make sure to add it to your action bar")
end
